src/core/unit.c

```diff
@@ -3291,21 +3291,59 @@ int unit_add_dependency(
 }
 
 int unit_add_two_dependencies(Unit *u, UnitDependency d, UnitDependency e, Unit *other, bool add_reference, UnitDependencyMask mask) {
-        int r, s;
+        int r = 0, s = 0;
 
         assert(u);
+        assert(d >= 0 || e >= 0);
 
-        r = unit_add_dependency(u, d, other, add_reference, mask);
-        if (r < 0)
-                return r;
+        if (d >= 0) {
+                r = unit_add_dependency(u, d, other, add_reference, mask);
+                if (r < 0)
+                        return r;
+        }
 
-        s = unit_add_dependency(u, e, other, add_reference, mask);
-        if (s < 0)
-                return s;
+        if (e >= 0) {
+                s = unit_add_dependency(u, e, other, add_reference, mask);
+                if (s < 0)
+                        return s;
+        }
 
         return r > 0 || s > 0;
 }
 
+int unit_add_dependencies_on_real_shutdown_targets(Unit *u) {
+        int r;
+
+        assert(u);
+
+        if (!u->survive)
+                return unit_add_two_dependencies_by_name(
+                                u,
+                                UNIT_BEFORE,
+                                UNIT_CONFLICTS,
+                                SPECIAL_SHUTDOWN_TARGET,
+                                /* add_reference= */ true,
+                                UNIT_DEPENDENCY_DEFAULT);
+
+        FOREACH_STRING(target,
+                       SPECIAL_REBOOT_TARGET,
+                       SPECIAL_KEXEC_TARGET,
+                       SPECIAL_HALT_TARGET,
+                       SPECIAL_POWEROFF_TARGET) {
```

Comment:
I am not convince this really is desirable like this. I think the xattr approach should be useful for storage services behind rootfs too, taking over the old argv[0][0] hack. But for those you definitely want them to survive well into poweroff/reboot too.

Reply:
nope. we manage cgroups, other people are not supposed to make changes to them, except when delegated.